[ceph-csi] Add migration from in-tree driver to csi driver (#2973)
deckhouse/deckhouse: modules/099-ceph-csi/docs/internal/INTREE_MIGRATION_RU.md

# Миграция с in-tree RBD драйвера на CSI (Ceph CSI)

> Для упрощения процесса миграции был написан скрипт [rbd-in-tree-to-ceph-csi-migration-helper.sh](../../tools/rbd-in-tree-to-ceph-csi-migration-helper.sh).
> Перед запуском необходимо удалить Pod (поскейлить в 0 StatefulSet/Deployment) использующий PVC. В процессе миграции будет необходимо вручную выполнить команду в ceph-кластере для переименования rbd-образа (Ceph CSI использует другой формат имени).
> 
> Во время работы скрипта будет сделан бэкап манифестов мигрируемых PVC и PV, далее они будут удалены и в итоге будут созданы новые PVC и PV. Удаление PV не приведет к удалению rbd-образа в ceph-кластере, т.к. перед удалением PV он будет переименован.

Reviewer: ```suggestion
Для упрощения процесса миграции был написан скрипт [rbd-in-tree-to-ceph-csi-migration-helper.sh](https://github.com/deckhouse/deckhouse/blob/main/modules/099-ceph-csi/tools/rbd-in-tree-to-ceph-csi-migration-helper.sh).
Перед запуском необходимо удалить Pod (поскейлить в ноль StatefulSet/Deployment) использующий PVC. В процессе миграции будет необходимо вручную выполнить команду в ceph-кластере для переименования RBD-образа (Ceph CSI использует другой формат имени).

Во время работы скрипта будет сделан бэкап манифестов мигрируемых PVC и PV, далее они будут удалены и в итоге будут созданы новые PVC и PV. Удаление PV не приведет к удалению RBD-образа в ceph-кластере, т.к. перед удалением PV он будет переименован.
```

  capacity:
    storage: 1Gi
  phase: Bound
```

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubernetes.io/createdby: rbd-dynamic-provisioner
    pv.kubernetes.io/bound-by-controller: "yes"
    pv.kubernetes.io/provisioned-by: kubernetes.io/rbd
  creationTimestamp: "2022-11-03T13:15:49Z"
  finalizers:
  - kubernetes.io/pv-protection
  name: pvc-cd6f7b26-d768-4cab-88a4-baca5b242cc5
  resourceVersion: "8956671"
  uid: 4ab7fcf4-e8db-426e-a7aa-f5380ef857c7
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: data-test-0
    namespace: default
    resourceVersion: "8956643"
    uid: cd6f7b26-d768-4cab-88a4-baca5b242cc5
  mountOptions:
  - discard
  persistentVolumeReclaimPolicy: Delete
  rbd:
    image: kubernetes-dynamic-pvc-f32fea79-d658-4ab1-967a-fb6e8f930dec
    keyring: /etc/ceph/keyring
    monitors:
    - 192.168.4.215:6789
    pool: kube
    secretRef:
      name: ceph-secret
    user: kube
  storageClassName: rbd
  volumeMode: Filesystem
status:
  phase: Bound
```

### Манифесты PVC и PV из которых будут заимствованы параметры характерные для Ceph CSI

> Используется StorageClass соданный модулем ceph-csi.

Reviewer: ```suggestion
Используется StorageClass созданный модулем ceph-csi.
```

  name: pvc-abdbb7ea-5da6-47f3-8b76-b968a93b7bc1
  resourceVersion: "8950562"
  uid: 6200ce15-b6f2-45af-94d0-828913e850d0
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: sample
    namespace: default
    resourceVersion: "8950550"
    uid: abdbb7ea-5da6-47f3-8b76-b968a93b7bc1
  csi:
    controllerExpandSecretRef:
      name: csi-new
      namespace: d8-ceph-csi
    driver: rbd.csi.ceph.com
    fsType: ext4
    nodeStageSecretRef:
      name: csi-new
      namespace: d8-ceph-csi
    volumeAttributes:
      clusterID: 60f356ee-7c2d-4556-81be-c24b34a30b2a
      imageFeatures: layering
      imageName: csi-vol-880ec27e-5b75-11ed-a252-fa163ee74632
      journalPool: kube
      pool: kube
      storage.kubernetes.io/csiProvisionerIdentity: 1666697721019-8081-rbd.csi.ceph.com
    volumeHandle: 0001-0024-60f356ee-7c2d-4556-81be-c24b34a30b2a-0000000000000005-880ec27e-5b75-11ed-a252-fa163ee74632
  mountOptions:
  - discard
  persistentVolumeReclaimPolicy: Delete
  storageClassName: ceph-csi-rbd
  volumeMode: Filesystem
status:
  phase: Bound
```

### Переименование RBD-образа в ceph-кластере

> Переименование необходимо, поскольку Ceph CSI драйвер использует другой формат имени rbd-образа.

Reviewer: ```suggestion
Переименование необходимо, поскольку Ceph CSI драйвер использует другой формат имени RBD-образа.
```

Команда выполняется в Ceph-кластере:

```shell
rbd mv kube/kubernetes-dynamic-pvc-<rbd-image-uid> kube/csi-vol-<rbd-image-uid>
```

> * `kube` - имя пула в ceph-кластере;
> * `kubernetes-dynamic-pvc-<uid>` - формат имени rbd-образа используемый in-tree драйвером;
> * `csi-vol-<uid>` - формат имени rbd-образа используемый Ceph CSI.

Reviewer: ```suggestion
* `kube` - имя пула в Ceph-кластере;
* `kubernetes-dynamic-pvc-<uid>` - формат имени RBD-образа используемый in-tree драйвером;
* `csi-vol-<uid>` - формат имени RBD-образа используемый Ceph CSI.
```

### Удаление PVC и PV из кластера

```bash
kubectl -n default delete pvc data-test-0
kubectl delete pv pvc-cd6f7b26-d768-4cab-88a4-baca5b242cc5
```

> Т.к. на предыдущем шаге мы переименовали rbd-образ в ceph-кластере, то удаление PersistentVolume не повлечет удаление образа.

Reviewer: ```suggestion
Т.к. на предыдущем шаге мы переименовали RBD-образ в Ceph-кластере, то удаление PersistentVolume не повлечет удаление образа.
```

  finalizers:
  - kubernetes.io/pv-protection
  name: pvc-cd6f7b26-d768-4cab-88a4-baca5b242cc5
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: data-test-0
    namespace: default
    uid: cd6f7b26-d768-4cab-88a4-baca5b242cc7
  mountOptions:
  - discard
  persistentVolumeReclaimPolicy: Delete
  csi:
    controllerExpandSecretRef:
      name: csi-new
      namespace: d8-ceph-csi
    driver: rbd.csi.ceph.com
    fsType: ext4
    nodeStageSecretRef:
      name: csi-new
      namespace: d8-ceph-csi
    volumeAttributes:
      clusterID: 60f356ee-7c2d-4556-81be-c24b34a30b2a
      imageFeatures: layering
      imageName: csi-vol-f32fea79-d658-4ab1-967a-fb6e8f930dec
      journalPool: kube
      pool: kube
      storage.kubernetes.io/csiProvisionerIdentity: 1666697721019-8081-rbd.csi.ceph.com
  volumeHandle: 0001-0024-60f356ee-7c2d-4556-81be-c24b34a30b2a-0000000000000005-f32fea79-d658-4ab1-967a-fb6e8f930dec
  storageClassName: ceph-csi-rbd
  volumeMode: Filesystem
```

Создадим объект в кластере используя этот манифест.

На этом миграция завершена.

> **Внимание!** Перед переходом на containerd убедитесь, что в кластере не используются сборщики логов, помимо log-shipper. Если такие имеются, то стоит либо от них отказаться в пользу модуля log-shipper, либо перенастроить их на работу с containerd. Containerd имеет другой формат логов и хранит файлы логов по другому пути.

Reviewer: ```suggestion
**Внимание!** Перед переходом на containerd убедитесь, что в кластере не используются сборщики логов, помимо log-shipper. Если такие имеются, то стоит либо от них отказаться в пользу модуля [log-shipper](https://deckhouse.io/documentation/v1/modules/460-log-shipper/), либо перенастроить их на работу с containerd. Containerd имеет другой формат логов и хранит файлы логов по другому пути.
```
